Gems/ScriptAutomation/Code/Source/ImGui/ImGuiAssetBrowser.cpp

```diff
@@ -0,0 +1,468 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <ImGui/ImGuiAssetBrowser.h>
+#include <AzCore/IO/SystemFile.h>
+#include <AzCore/std/sort.h>
+#include <AzCore/Serialization/Utils.h>
+#include <ScriptableImGui.h>
+
+namespace ScriptAutomation
+{
+    void ImGuiAssetBrowser::Reflect(AZ::ReflectContext* context)
+    {
+        ImGuiAssetBrowser::ConfigFile::Reflect(context);
+    }
+
+    void ImGuiAssetBrowser::ConfigFile::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<ConfigFile>()
+                ->Version(0)
+                ->Field("PinnedAssetPaths", &ConfigFile::m_pinnedAssetPaths)
+                ->Field("ExpandRoot", &ConfigFile::m_expandRoot)
+                ->Field("ExpandAvailableList", &ConfigFile::m_expandAvailableList)
+                ->Field("ExpandPinnedList", &ConfigFile::m_expandPinnedList)
+                ;
+        }
+    }
+
+    ImGuiAssetBrowser::ImGuiAssetBrowser(AZStd::string_view configFilePath)
+    {
+        m_configFilePath = configFilePath;
+    }
+
+    void ImGuiAssetBrowser::Activate()
+    {
+        // The original m_configFilePath passed to the constructor likely starts with "@user@" which needs to be replaced with the real path.
+        // The constructor is too early to resolve the path so we do it here on activation.
+        char configFileFullPath[AZ_MAX_PATH_LEN] = {0};
+        AZ::IO::FileIOBase::GetInstance()->ResolvePath(m_configFilePath.c_str(), configFileFullPath, AZ_MAX_PATH_LEN);
+        m_configFilePath = configFileFullPath;
+
+        AzFramework::AssetCatalogEventBus::Handler::BusConnect();
+    }
+
+    void ImGuiAssetBrowser::Deactivate()
+    {
+        if (!m_configFilePath.empty())
+        {
+            // We only report this message in Deactivate(), not inside SaveConfigFile(), to avoid spamming
+            // this message when SaveConfigFile() is called in OnTick()
+            AZ_TracePrintf("ImGuiAssetBrowser", "Saved settings to '%s'\n", m_configFilePath.c_str());
+
+            SaveConfigFile();
+        }
+
+        AzFramework::AssetCatalogEventBus::Handler::BusDisconnect();
+    }
+
+    void ImGuiAssetBrowser::OnCatalogAssetAdded(const AZ::Data::AssetId& assetId)
+    {
+        OnCatalogChanged(assetId);
+    }
+        
+    void ImGuiAssetBrowser::OnCatalogAssetRemoved(const AZ::Data::AssetId& assetId, const AZ::Data::AssetInfo&)
+    {
+        OnCatalogChanged(assetId);
+    }
+
+    void ImGuiAssetBrowser::OnCatalogChanged(const AZ::Data::AssetId& assetId)
+    {
+        AZ::Data::AssetInfo assetInfo;
+        AZ::Data::AssetCatalogRequestBus::BroadcastResult(assetInfo, &AZ::Data::AssetCatalogRequestBus::Events::GetAssetInfoById, assetId);
+
+        if (m_includedAssetFilter && m_includedAssetFilter(assetInfo))
+        {
+            m_needsRefresh = true;
+        }
```

Comment:
This doesn't set needs refresh if no asset filter is supplied, is that intentional?